paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/FlinkExternalCatalog.java

```diff
@@ -0,0 +1,488 @@
+import org.apache.flink.table.catalog.exceptions.PartitionSpecInvalidException;
+import org.apache.flink.table.catalog.exceptions.TableAlreadyExistException;
+import org.apache.flink.table.catalog.exceptions.TableNotExistException;
+import org.apache.flink.table.catalog.exceptions.TableNotPartitionedException;
+import org.apache.flink.table.catalog.exceptions.TablePartitionedException;
+import org.apache.flink.table.catalog.stats.CatalogColumnStatistics;
+import org.apache.flink.table.catalog.stats.CatalogTableStatistics;
+import org.apache.flink.table.descriptors.DescriptorProperties;
+import org.apache.flink.table.expressions.Expression;
+
+import java.io.IOException;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+import static org.apache.flink.table.factories.FactoryUtil.CONNECTOR;
+
+/** Catalog for persisting external tables. */
+public class FlinkExternalCatalog extends AbstractCatalog {
+
+    private final FlinkCatalog paimon;
+    private final String warehousePath;
+
+    private static final String EXTERNAL_TABLE_STORE_PATH = "TABLE";
+    private static final String EXTERNAL_FUNCTION_STORE_PATH = "FUNCTION";
+    private final FileIO fileIO;
+    private final String systemDbDir;
+
+    private final String externalFunctionDir;
+
+    private final String externalTableDir;
+
+    public FlinkExternalCatalog(FlinkCatalog paimon, FileIO fileIO, String warehousePath) {
+        super(paimon.getName(), paimon.getDefaultDatabase());
+        this.paimon = paimon;
+        this.fileIO = fileIO;
+        this.warehousePath = warehousePath;
+        this.systemDbDir =
+                this.warehousePath + Path.SEPARATOR + "FLINK_EXTERNAL_METADATA" + Path.SEPARATOR;
+        externalFunctionDir = systemDbDir + EXTERNAL_FUNCTION_STORE_PATH + Path.SEPARATOR;
+        externalTableDir = systemDbDir + EXTERNAL_TABLE_STORE_PATH + Path.SEPARATOR;
+        try {
+            fileIO.mkdirs(new Path(externalTableDir));
+        } catch (IOException ignore) {
+
+        }
+        try {
+            fileIO.mkdirs(new Path(externalFunctionDir));
+        } catch (IOException ignore) {
+
+        }
+    }
+
+    public org.apache.paimon.catalog.Catalog catalog() {
+        return paimon.catalog();
+    }
+
+    private String externalTableSchemaDir(ObjectPath table) {
+        return externalTableDir
+                + Path.SEPARATOR
+                + table.getDatabaseName()
+                + Path.SEPARATOR
+                + table.getObjectName()
+                + Path.SEPARATOR;
+    }
+
+    @Override
+    public void open() throws CatalogException {
+        paimon.open();
+    }
+
+    @Override
+    public void close() throws CatalogException {
+        paimon.close();
+    }
+
+    @Override
+    public List<String> listDatabases() throws CatalogException {
+        return paimon.listDatabases();
```

Comment:
Do we need to list databases for external tables?

Reply:
I think saving different external tables to different databases would give the user a better experience. For example
![d6e7c95bf3ca4c629c2c634a9099135](https://github.com/apache/incubator-paimon/assets/35210666/b5e7ed7a-0dba-41cb-8c8c-85d0d6eaa41b)

Reply:
I cannot get your point.
I mean list databases for external tables. There is only paimon databases.

Reply:
I made the external table use the same databases as the paimon table.